globus_cli/commands/session/update.py

```diff
@@ -1,25 +1,86 @@
-import uuid
-
 import click
 
 from globus_cli.helpers import (
     do_link_auth_flow,
     do_local_server_auth_flow,
     is_remote_session,
 )
-from globus_cli.parsing import command, no_local_server_option
+from globus_cli.parsing import IdentityType, command, no_local_server_option
 from globus_cli.services.auth import get_auth_client
 
 
+def _identity_set(auth_client):
+    res = auth_client.oauth2_userinfo()
+    try:
+        return res["identity_set"]
+    except KeyError:
+        click.echo(
+            "Your current login does not have the consents required "
+            "to view your full identity set. Please log in again "
+            "to agree to the required consents.",
+            err=True,
+        )
+        click.get_current_context().exit(1)
+
+
+def _update_session_params_all_case(auth_client, session_params):
+    """if --all use every identity id in the user's identity set"""
+    identity_ids = [x["sub"] for x in _identity_set(auth_client)]
+    # set session params once we have all identity ids
+    session_params["session_required_identities"] = ",".join(identity_ids)
+
+
+def _update_session_params_identities_case(auth_client, session_params, identities):
+    """
+    given a set of identities (which must be either a mix of usernames and IDs or a list
+    of domains), use that to update the session as appropriate
+    """
+    identity_ids = [i.value for i in identities if i.idtype == "identity"]
+    identity_usernames = [i.value for i in identities if i.idtype == "username"]
+    identity_domains = [i.value for i in identities if i.idtype == "domain"]
+
+    if identity_domains and (identity_ids or identity_usernames):
+        raise click.UsageError(
+            "domain-type identities and user-type identities are mutually exclusive"
+        )
+
+    # if usernames were used, fetch the identity set and pull identity IDs from there
+    # do not use 'get_identities' as it will easily return identities which are not in
+    # your identity set
+    if identity_usernames:
+        identity_set = _identity_set(auth_client)
+
+        for name in identity_usernames:
+            for identity in identity_set:
+                if identity["username"] == name:
+                    identity_ids.append(identity["sub"])
+                    break
+            else:
+                click.echo("'{}' is not in your identity set".format(name), err=True)
+                click.get_current_context().exit(1)
+
+    # update session params once we have resolved usernames (if necessary)
+    if identity_ids:
+        session_params["session_required_identities"] = ",".join(identity_ids)
```

Comment:
This allows for the possibility of identity IDs which aren't in the user's identity set.

I've just checked, and attempting to authorize with an identity not in your set leads to an error. It's better to error out early with a clear message from the CLI in this case.

I think I should add such handling, which might allow for restructuring to fetch the identity set at an earlier step.